addons/l10n_it_edi_withholding/models/account_move.py

```diff
@@ -186,21 +189,57 @@ def _l10n_it_edi_get_extra_info(self, company, document_type, body_tree, incomin
                  + type_tax_use_domain),
                 vat_only=False)
             if pension_fund_tax:
-                pension_fund_taxes.append(pension_fund_tax)
+                pension_fund_taxes[vat_tax_factor_percent] = pension_fund_tax
             else:
                 message_to_log.append(Markup("%s<br/>%s") % (
                     _("Pension Fund tax not found"),
                     self.env['account.move']._compose_info_message(body_tree, '.'),
                 ))
         extra_info["pension_fund_taxes"] = pension_fund_taxes
 
+        # If the AssoSoftware specs are used on the invoice, then only apply
+        # the Pension Fund tax to the lines that show an AswCassPre
+        # additional tag (AltriDatiGestionali)
+        selector = ".//AltriDatiGestionali/TipoDato[contains(text(), 'AswCassPre')]"
+        if self.get_tag(body_tree, selector) is not None:
+            extra_info["pension_fund_assosoftware_tags"] = True
+
         return extra_info, message_to_log
 
+    def get_tag(self, element, selector):
+        if element is None:
+            return None
+        sub = element.xpath(selector)
+        return sub[0] if (sub is not None and sub) else None
+
+    def _get_pension_fund_tax_for_line(self, element, extra_info):
+        """ If there are AssoSoftware specific AltriDatiGestionale 'AswCassPre'
+            tags that specify which lines have pension funds, use them.
+            Otherwise apply the pension fund on all lines that have the related AliquotaIVA
+        """
+        pension_fund_map = extra_info.get('pension_fund_taxes', {})
+        if pension_fund_map and extra_info.get('pension_fund_assosoftware_tags'):
+            selector = ".//AltriDatiGestionali[TipoDato[contains(text(),'AswCassPre')]]/RiferimentoTesto"
```

Comment:
Do you really want `/RiferimentoTesto`?
If I'm not mistaken, this will only take `RiferimentoTesto` if it's on the root of the document :thinking:

Reply:
... no?

`.` from this point
`//` after whatever level
`AltriDatiGestionali`  find tags with this name
`[TipoDato`  that have this child
`[contains(text(),'AswCassPre')]]` that contain this text
`/RiferimentoTesto` ... and select this child.

Looks like it works and it saves some useless Python mangling

![image](https://github.com/user-attachments/assets/cf1c8e39-c993-4ceb-8ee3-1e2c7f90288f)